app/rtc_compare.py

```diff
@@ -0,0 +1,329 @@
+            val_1 = np.array(val_1)
+
+        if not isinstance(val_2,np.ndarray):
+            val_1 = np.array(val_2)
+
+    else:
+        val_1 = np.array(hdf5_obj_1[str_key])
+        val_2 = np.array(hdf5_obj_2[str_key])
+
+    shape_val_1 = np.array(val_1).shape
+    shape_val_2 = np.array(val_2).shape
+
+    if not (shape_val_1 == shape_val_2):
+        # Dataset or attribute shape does not match
+        return False
+
+    elif len(shape_val_1)==0 and len(shape_val_2)==0:
+        # Scalar value
+        if val_1.dtype=='float' and val_2.dtype=='float':
+            return np.array_equal(val_1, val_2, equal_nan=True)
+        else:
+            return np.array_equal(val_1, val_2)
+
+    elif len(shape_val_1)==1 and len(shape_val_2)==1:
+        # 1d vector
+        if 'shape' in dir(val_1[0]):
+            if isinstance(val_1[0], np.void):
+                list_val_1 = list(itertools.chain.from_iterable(val_1))
+                val_1_new = [None] * len(list_val_1)
+                for i_val, element_1 in enumerate(list_val_1):
+                    if isinstance(element_1, h5py.h5r.Reference):
+                        val_1_new[i_val] = hdf5_obj_1[element_1]
+                    else:
+                        val_1_new[i_val] = element_1
+
+                val_1 = val_1_new
+
+            elif (len(val_1[0].shape) == 1) and (isinstance(val_1[0][0], h5py.h5r.Reference)):
+                list_val_1 = list(itertools.chain.from_iterable(val_1))
+                val_1_new=[None] * len(list_val_1)
+                for i_val, element_1 in enumerate(list_val_1):
+                    if isinstance(element_1, h5py.h5r.Reference):
+                        val_1_new[i_val] = hdf5_obj_1[element_1]
+                    else:
+                        val_1_new[i_val] = element_1
+                val_1 = val_1_new
+
+        if 'shape' in dir(val_2[0]):
+            if isinstance(val_2[0], np.void):
+                list_val_2 = list(itertools.chain.from_iterable(val_2))
+                val_2_new = [None] * len(list_val_2)
+                for i_val, element_2 in enumerate(list_val_2):
+                    if isinstance(element_2, h5py.h5r.Reference):
+                        val_2_new[i_val] = hdf5_obj_2[element_2]
+                    else:
+                        val_2_new[i_val] = element_2
+
+                val_2 = val_2_new
+
+            elif (len(val_2[0].shape) == 1) and (isinstance(val_2[0][0], h5py.h5r.Reference)):
+                list_val_2 = list(itertools.chain.from_iterable(val_2))
+                val_2_new = [None]*len(list_val_2)
+                for i_val, element_2 in enumerate(list_val_2):
+                    if isinstance(element_2, h5py.h5r.Reference):
+                        val_2_new[i_val] = hdf5_obj_2[element_2]
+                    else:
+                        val_2_new[i_val] = element_2
+
+                val_2 = val_2_new
+
+        if isinstance(val_1, list) and isinstance(val_2, list):
```

Comment:
I think you can check earlier if both `val_1` and `val_2` are the same class

Reply:
The reason to check the instance type in that line is to differentiate the comparison strategy, rather than to check if they are identical.
By default, `val_1` and `val_2` are `np.ndarray`. In case they contain HDF5 object reference (i.e. `h5py.h5r.Reference`), the referenced data replaces the object reference, and `val_1` `val_2`are converted to `list`.

Two reasons for this internal behavior:
- The referenced objects are often vectors and arrays. It is more appropriate to compare `val_1` and `val_2` using sth. like `np.allclose()` element by element
- The composition of the dereferenced list can be complex i.e. multiple datatypes of elements. Example below is for attribute `REFERENCE_LIST` in  `/science/CSAR/RTC/grids/frequencyA/yCoordinates`
>print(val_1)
[<HDF5 dataset "VV": shape (1418, 3497), type "<f4">, 0, <HDF5 dataset "VH": shape (1418, 3497), type "<f4">, 0, <HDF5 dataset "numberOfLooks": shape (1418, 3497), type "<f4">, 0, <HDF5 dataset "areaNormalizationFactor": shape (1418, 3497), type "<f4">, 0, <HDF5 dataset "incidenceAngle": shape (1418, 3497), type "<f4">, 0, <HDF5 dataset "localIncidenceAngle": shape (1418, 3497), type "<f4">, 0]
- Converting the lists like `val_1/2` are not favorable operation:
>np.array(val_2)
<string>:1: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.`

Reply:
Thank you for this info, @seongsujeong . Is it possible to know the path to the referenced dataset? Maybe you can just check if they point to the same dataset.

Reply:
@gshiroma We can't do that directly from the object, but what we can do is to dereference it and access `.name` of the data. I've made the change as the snippet below:

```python
val_2_new[i_val] = np.str_(hdf5_obj_2[element_2].name)
```

Reply:
Well, but you'll are comparing all elements of the HDF5 dataset. Soon or later that element (that was referenced by another one) will be checked as well.

Reply:
@gshiroma Yes, and I personally think that is more desirable behavior for the object reference.